src/data/dataset.py

```diff
@@ -19,88 +49,149 @@ class Dataset(object):
+                        dirs.append(artist)
+
             num_dirs = len(dirs)
-            last_log = 0
+
             for artist_index, artist in enumerate(dirs):
-                # log progress every second
-                if time.time() - last_log >= 1:
-                    log.info("Preprocessing data. %s%%" % int(100*artist_index/num_dirs))
-                    last_log = time.time()
                 songs = os.listdir(os.path.join(root, artist))
-                if len(songs) >= min_songs:
-                    artists.append(artist)
+                # We only want .txt and .mid files. Filter all others.
+                songs = [song for song in songs if loader.is_song(song)]
+                # populate `valid_songs[artist]`
+                if validate:
+                    # log progress at most every second
+                    if time.time() - last_log >= 1:
+                        if last_log_percent != '%.2f' % (100*artist_index/num_dirs):
+                            last_log_percent = '%.2f' % (100*artist_index/num_dirs)
+                            log.info("Preprocessing data. %s%%" % last_log_percent)
+                            last_log = time.time()
+                    if artist not in valid_songs:
+                        valid_songs[artist] = set()
+                    songs_to_validate = [song for song in songs if song not in valid_songs[artist]]
+                    song_files = [os.path.join(root, artist, song) for song in songs_to_validate]
+                    if parallel:
+                        mapped = pool.map(loader.validate, song_files)
+                    else:
+                        mapped = map(loader.validate, song_files)
+                    validated = itertools.compress(songs_to_validate, mapped)
+                    for song in validated:
+                        song_file = os.path.join(root, artist, song)
+                        if persist:
+                            self.metadata.write('valid_songs.csv', '%s,%s\n' % (artist, song))
+                        valid_songs[artist].add(song)
+                else:
+                    valid_songs[artist] = set(songs)
+
+                if len(valid_songs[artist]) >= min_songs:
+                    all_artists.append(artist)
                 else:
                     skipped_count += 1
+            pool.close()
+            pool.join()
             if skipped_count > 0:
                 log.info("%s artists don't have K+K'=%s songs. Using %s artists" % (
-                    skipped_count, min_songs, len(artists)))
-            train_count = int(float(split_proportions[0]) / sum(split_proportions) * len(artists))
-            val_count = int(float(split_proportions[1]) / sum(split_proportions) * len(artists))
-            np.random.shuffle(artists)
+                    skipped_count, min_songs, len(all_artists)))
+            train_count = int(float(split_proportions[0]) / sum(split_proportions) * len(all_artists))
```

Comment:
this is my first time looking at this code, so apologies if i'm misunderstanding the way the code is structured, but i'm a little confused with the way `artists_in_split` is built.
it seems like Dataset is called 3 times (once with `split == train`, once with `split == val`, and once with `split == eval`. that means that the `np.random.shuffle` will be run 3 times, which means that you will likely end up with the same artist showing up in different splits, and some artists not showing up in any split.
is this the intended behaviour?

Reply:
Yeah. That will happen if you're not using persistence. When I initially wrote this, I didn't think users would care if the splits were overlapping if they weren't saving the splits, but maybe that's not the case. I could add some extra code to the metadata handler to handle non-persisted splits as you'd expect.

Reply:
cool. you could put this as a TODO for now if you want.

Reply:
SGTM